### ISSUE TITLE
Allow address re-use

### ISSUE THREAD

**Issue by apbendi** [ScopeLift/umbra-protocol#56]
Add to token balance instead of setting a flag. This would avoid the need to set a flag, reducing our gas costs.

**Comment by apbendi** [comment]
Currently we have a mapping:

`mapping(address => TokenPayment) tokenPayments;`

Where `address` is the stealth address that received the payment, i.e. the "acceptor" in the code's parlance, and `TokenPayment` is:

```solidity
struct TokenPayment {
    address token;
    uint256 amount;
  }
```

Our withdraw method signature is `function withdrawToken(address _acceptor)`, and we use the above mapping to look up the token address and amount the acceptor must be withdrawing.

If we allow reuse, the mapping between address and `TokenPayment` will not be sufficient, because someone could send a payment of a different kind of token. The current system has no way to account for this.

#### Options:

1. Change to a mapping that accounts for different tokens, for example `mapping(address => mapping(address => amount)) tokenPayments;`. In this mapping, the first address is the acceptor, the second address is the token contract address, and the amount is the amount. If we do this, we'll have to change our withdraw signature to `function withdrawToken(address _acceptor, address _token)`, namely the user calling `withdrawToken` will have to send the token address.

2. Continue preventing address re-use, but instead of using the extra flag, simply check the existing `tokenPayments` entry for the acceptor address for a non-zero address or amount. This avoids setting the extra flag, which saves the gas, which was the goal of this change. It also avoids the existing payment from being overwritten by a future payment, which was the bug we were trying to avoid when we introduced the flag. Finally, it would still technically allow a stealth address to be re-used, but only after the receiver has withdrawn it (thus clearing the mapping entry).

**Comment by apbendi** [comment]
closed in #64
